developer_docs/semantic_stage.md

```diff
@@ -0,0 +1,184 @@
+
+## Impose Restrictions
+
+Not all valid Python code can be translated by Pyccel.
+This is for one of three reasons:
+1.  Support has not yet been added but is planned (e.g. classes)
+
+2.  It would be impractical to support the code in a low-level language. E.g. functions which return objects of different types :
+    ```python
+    def f(a : bool, b : bool):
+        if a and b:
+            return 1.0, False
+        elif b:
+            return 10, 4j
+        elif a:
+            return True
+        else:
+            return
+    ```
+
+3.  A conscious choice has been made to not support code as there is no way to obtain a performant translation.
+    Pyccel is designed to handle HPC applications.
+    Therefore if the code cannot be performant, we prefer to warn the user so they can fix it rather than generating slow code (e.g. inhomogeneous lists).
+
+When the translation fails due to Pyccel restrictions it is important to raise a clear error.
+Users may be surprised and annoyed to find that their functioning Python code will not translate.
+In order to facilitate their interactions with Pyccel, we therefore want to make sure that errors fully explain the problems with their code, and point towards any potential documentation which may show them how to fix the code.
+
+Error handling uses the classes in the file [pyccel.errors.errors](../pyccel/errors/errors.py).
+Errors in the semantic stage should raise a `PyccelSemanticError`.
+Where possible this should be done by accessing the `Errors()` singleton and calling the `report` function.
+This function takes several arguments (see docstring for more details).
+The most important arguments are:
+-   _message_ : Describe the issue that lead to the error
+
+-   _symbol_ : The Python AST object should be passed here. This object contains information about its position in the file (line number, column) which ensures the user can more easily locate their error
+
+-   _severity_ : The severity level must be one of the following:
+    -   _warning_ : An error will be printed but Pyccel will continue executing
+    -   _error_ : An error will be printed but Pyccel will continue executing the syntactic stage
+    -   _fatal_ : An error will be printed and Pyccel will stop executing. This is the level used the most frequently in the semantic stage as, if information such as the type cannot be determined, the object cannot be created which will cause problems later in the execution.
+
+It should be noted that Pyccel assumes that the user has provided valid code.
+It is not feasible to provide error messages for every possible coding error so we limit ourselves to simple validity checks and Pyccel restriction errors.
+
+## Function Recognition
+
+Function calls are split into two groups which are handled in different ways.
```

Comment:
After the syntactic stage, are all function calls represented as `FunctionCall` nodes?

Reply:
Yes

Reply:
Then I think it is worth emphasizing this!

The `expr` node is always a `FunctionCall` object, and the corresponding function definition is always contained in `expr.funcdef`. Depending on the type of this attribute, the call is handled in different ways by `visit_FunctionCall`.

Reply:
Having double-checked this it seems that there is one exception (`print`). I have described this here : 6443555d
I suspect the different treatment of `print` comes from a time when python 2 was parsed but I'm not sure. I don't see any reason for the different treatment at the moment. It would be better to open a small PR to remove this special treatment

Reply:
You are correct, originally Pyccel wanted to support Python 2 and 3. I agree with your suggestion: Python 3 `print` is a built-in function like all the others, so we should remove this special treatment.